lib/sla.js

```diff
@@ -45,80 +48,135 @@ function designateMixingBridge(client, bridgeName) {
 }
 
 /**
- * Adds inbound and originated channels to the bridge.
+ * Updates the state of the shared extension.
  * @param {Object} client - Object that contains information from the ARI
  *   connection.
- * @param {Object} inbound - Incoming channel to be added to the bridge.
+ * @param {String} name - name of the shared extension to update.
+ * @param {String} state - state to update the shared extension to.
+ */
+function updateState(client, name, state) {
+  var deviceState = Q.denodeify(client.deviceStates.update.bind(client));
+  deviceState({deviceName: 'Stasis:' + name, deviceState: state});
+}
+
+/**
+ * Accesses the current state of the shared extension.
+ * @param {Object} client - Object that contains information from the ARI
+ *   connection.  Houses the device states
+ * @param {String} name - name of the shared extension to access.
+ * @return {String} - the current device state of this extension
+ */
+function getState(client, name) {
+  var getDeviceState = Q.denodeify(client.deviceStates.get.bind(client
+        .deviceStates));
+  return getDeviceState({deviceName: 'Stasis:' + name}).then(function (ds) {
+    return ds.state;
+  });
+}
+
+/**
+ * Adds the caller channel and originated channel (if applicable) to the bridge.
+ * @param {Object} client - the client (used for hanging up at the end).
+ * @param {Object} channel - Incoming channel to be added to the bridge.
  * @param {Object} dialed - Created/dialed channel to be added as well
  * @param {Object} bridge - Bridge that channels are to be added to.
+ * @param {Object} extension - the object data related to this sharedExtension
  * @return {Q} Q promise object.
  */
-function addChannelsToBridge(client, inbound, dialed, bridge) {
+function addChannelsToBridge(client, channel, dialed, bridge, extension) {
+  // The only way that the channel length will be 1 is if the outside user is
+  // still inside of the shared extension waiting for others to answer.
+  if (bridge.channels.length === 1) {
```

Comment:
I don't agree with that assumption. You should not expect that you only have a station left in the shared extension.

Consider the following scenarios:
* Station joins the SE (channels == 1)
* Station calls a trunk (channels == 2)
* Station and trunk talk. Station hits hold, causing them to leave the bridge (channels == 1)
* trunk stays in SE, and hears hold music.

Another scenario:
* Station 1 joins the SE (channels == 1)
* Station 2 joins the SE (channels == 2)
* Trunk joins the SE (channels == 3)
* Trunk hangs up (channels == 2)
* Station 1 hits hold, disconnecting
* Stations 2 does _not_ hear MoH

Both of these scenarios are mentioned in the issue requirements.

RIght now, you have nothing that I can find that keeps track of the types of channels in the SE, and makes use of that knowledge when the channel count changes. I would expect to see some data structure holding the channels and their types in the bridge, and when the state changes, alters the state of the bridge accordingly based on that information.